### ISSUE TITLE
Encoding error on Windows I/O with Python 3.6

### ISSUE THREAD

**Issue by JGoutin** [h5py/h5py#818]
I use H5py 2.7rc2 with Python 3.6.0 64bits on Windows 7 64bits.

In this example, I try to write and read two files named `test_e.h5` and `test_é.h5` : 

Theses problems are new (Only since I updated to Python 3.6).

Maybe linked to [PEP529](https://docs.python.org/3/whatsnew/3.6.html#pep-529-change-windows-filesystem-encoding-to-utf-8) change.

#### Write test : 

```python
f = h5py.File('test_e.h5', 'w')
f.close()
f = h5py.File('test_é.h5', 'w')
f.close()
print('Directory content after test:', os.listdir())
```
Result : 
```python
Directory content after test:  ['test_e.h5', 'test_Ã©.h5']
```
Bad file name: `test_Ã©.h5` in place of `test_é.h5`

#### Read Test

```python
print('Directory content before test: ', os.listdir())
f = h5py.File('test_e.h5', 'r')
f.close()
f = h5py.File('test_é.h5', 'r')
f.close()
```

Result :
```python
Directory content before test:  ['test_e.h5', 'test_é.h5']

Traceback (most recent call last):

  File "D:/Dev/format_Hdf5.py", line 16, in <module>
    f = h5py.File('test_é.h5', 'r')

  File "d:\app\python36\lib\site-packages\h5py\_hl\files.py", line 271, in __init__
    fid = make_fid(name, mode, userblock_size, fapl, swmr=swmr)

  File "d:\app\python36\lib\site-packages\h5py\_hl\files.py", line 101, in make_fid
    fid = h5f.open(name, flags, fapl=fapl)

  File "h5py\_objects.pyx", line 54, in h5py._objects.with_phil.wrapper (D:\Build\h5py\h5py-2.7.0rc2\h5py\_objects.c:2853)

  File "h5py\_objects.pyx", line 55, in h5py._objects.with_phil.wrapper (D:\Build\h5py\h5py-2.7.0rc2\h5py\_objects.c:2811)

  File "h5py\h5f.pyx", line 78, in h5py.h5f.open (D:\Build\h5py\h5py-2.7.0rc2\h5py\h5f.c:2130)

OSError: Unable to open file (Unable to open file: name = 'test_é.h5', errno = 2, error message = 'no such file or directory', flags = 0, o_flags = 0)
```
Existing file `test_é.h5` not found.

**Comment by andreabedini** [comment]
That's funny, we use `os.fsencode` https://github.com/h5py/h5py/blob/e611b7ca47e49d83e908d0368d2d9ced224a9de0/h5py/_hl/files.py#L268 to encode the path down to bytes.

**Comment by aragilar** [comment]
Yeah, this is PEP529 (mbcs is replaced by utf8 as default system encoding on Windows) where latin1 is the current ANSI codepage (this should mean we get something else when the locale is changed):
```
>>> "test_é.h5".encode("utf8").decode("latin1")
'test_Ã©.h5'
```
The real solution would be for HDF5 to fix their Windows path code to use the correct APIs (i.e. use the unicode APIs, rather than the ANSI ones, see https://tschoonj.github.io/blog/2014/11/06/hdf5-on-windows-utf-8-filenames-support/, https://lists.hdfgroup.org/pipermail/hdf-forum_lists.hdfgroup.org/2016-July/009625.html, https://lists.hdfgroup.org/pipermail/hdf-forum_lists.hdfgroup.org/2014-August/007988.html, https://lists.hdfgroup.org/pipermail/hdf-forum_lists.hdfgroup.org/2009-June/000197.html), but I don't see that happening any time soon (and doesn't fix existing code).

We could just patch HDF5 in our Windows releases, which fixes a whole bunch of problems (conda has a request to do this at https://github.com/conda-forge/hdf5-feedstock/issues/47, so we'd not be the only ones to do this).

Or we can try to work around this when we encode/decode (and hope we don't run into https://www.python.org/dev/peps/pep-0529/#id21).

I think patching HDF5 is the better alternative, as while I could write a patch to work around 3.6, there's a whole bunch of corner cases I'd have to worry about (we would have to check what bytes are being passed to filename functions, and decode/re-encode them, which could break code assuming bytes are being passed through without changes...). Also, patching HDF5 fixes the long term problem about having pathname which are not in the current codepage being unable to be accessed.

@JGoutin A workaround is to encode the filename (with `filename.encode('mbcs')`) before passing it to h5py, you need this for older versions of h5py on python 3.6 as well unless I'm very much mistaken.

Opinions on our options @cgohlke @jakirkham?

**Comment by andreabedini** [comment]
Great work @aragilar. I have exactly zero-knowledge of Windows APIs so I was a bit lost. Both PyTables and h5py now vendor HDF5 in their wheels, so we could use a patched version of HDF5. HDF5 is now using git (although on bitbucket) so we could have a mirror somewhere here on GitHub for our use.

@epourmal can you find in the bug tracker the issue Quincey refers to in https://lists.hdfgroup.org/pipermail/hdf-forum_lists.hdfgroup.org/2009-June/000197.html ? can you check with your team if there's a way to address this issue without a patch?

**Comment by JGoutin** [comment]
@aragilar using `filename.encode('mbcs')` work well as temporary fix for my code. Thanks.

Strangely, using `os.fsencode(filename)` even with filesystem forced to mbcs don't work :
```python
import sys
import os

# Force the use of 'mbcs' like versions of Python prior to 3.6.
sys._enablelegacywindowsfsencoding()

# Show actual file system encoding
encoding = sys.getfilesystemencoding()
print('Filesystem encoding:', encoding)

# os.fsencode(filename) VS filename.encode('mbcs')
print(os.fsencode('test_é.h5'), 'test_é.h5'.encode(encoding))
```
Result : 
```python
Filesystem encoding: mbcs
b'test_\xc3\xa9.h5' b'test_\xe9.h5'
```
Filesystem encoding is mbcs. But, encoded bytes are different.

**Comment by aragilar** [comment]
@JGoutin See https://hg.python.org/cpython/file/v3.6.0/Lib/os.py#l788, `os.fsencode` caches the encoding, which makes sense, as `sys._enablelegacywindowsfsencoding()` is new and unexpected, and generally not what you want, since it's global and may break other code, whereas using `PYTHONLEGACYWINDOWSFSENCODING` makes it a user decision, and hence the user knows about potential problems.

**Comment by JGoutin** [comment]
@aragilar As written in the doc, using [sys._enablelegacywindowsfsencoding()](https://docs.python.org/3.6/library/sys.html#sys._enablelegacywindowsfsencoding) or [PYTHONLEGACYWINDOWSFSENCODING](https://docs.python.org/3.6/using/cmdline.html#envvar-PYTHONLEGACYWINDOWSFSENCODING), should have the same behaviour.

Yes, using `sys._enablelegacywindowsfsencoding()` is not an options for fix the problem in h5py, or any other library. In my case, I only tried to use it as temporary fix in my application (And avoid to change all `h5py.File` calls in my code).

**Comment by tacaswell** [comment]
Re-milestoned as it does not look like there is any action we can take in h5py to address this.

A cleaned up version of this should end up in the docs.

**Comment by JGoutin** [comment]
A possible fix may be to not use `os.fsencode(filename)`/`os.fsdecode(filename)` if running on Python > 3.6 and Windows, but `filename.encode('mbcs')`/`filename.decode('mbcs')` in this case.

**Comment by tacaswell** [comment]
@JGoutin Can you put in a PR with that change?  I assume you have a system you can test this on?

**Comment by JGoutin** [comment]
@tacaswell OK, I'll do this PR. I have system for test, this can also be done with AppVeyor.

**Comment by aragilar** [comment]
@JGoutin If you haven't already created the patch, could you do the fix via creating two new functions, one which encodes filenames (which uses fspath and fsencode, plus your change), and one which decodes filenames, rather than modifying _fscodec? I was going to wait on 2.7.1 to fix this so as not to slow down the release (plus apply the hdf5 patch and add 3.6 to CI tests), but if you've got something ready now go ahead and make a PR.

**Comment by JGoutin** [comment]
@aragilar Is this (#818) OK for you ? I still replace fsencode/fsdecode function for avoid to replace them everywhere, but I didn't changed already fixed version from this file.

**Comment by aragilar** [comment]
@JGoutin See comments on #836.

I've created a separate issue to track the problems of unicode paths on Windows (#839, which is definitely not going to be resolved in 2.7), so this issue can just be the workaround to make Python 3.6 behave like previous versions.

**Comment by aragilar** [comment]
@JGoutin I've gone ahead and made a simpler version at #841, which includes documentation about what users should pass. Does it seem ok to you?
@tacaswell Are #838 and #841 fine to add to the 2.7 release, they should close this.

**Comment by JGoutin** [comment]
@aragilar it's OK for me 👍

**Comment by aragilar** [comment]
Fixed by #841.

**Comment by tschoonj** [comment]
I have been in touch with HDF support and they assured me that they will start using UTF-8 filenames on Windows starting with 1.10.3, which is due for release at the end of this year. 

That should settle this issue once and for all...

**Comment by tadeu** [comment]
HDF5 latest version is 1.10.4, and there still seems to be no support for UTF-8 filenames:

https://bitbucket.hdfgroup.org/projects/HDFFV/repos/hdf5/browse/src/H5win32defs.h#57

still using `_open` instead of  `_wopen` in Windows :/

**Comment by tschoonj** [comment]
I know... I asked them last week about this and all they replied the following:

> I checked the bug report, and the developer had added comments. The issue
is still open.
> 
> A concern the developer has is that the patch works around deficits in HDF5
such as a lack of proper UTF-8 support in the Windows POSIX layer. He thinks
that implementing the patch may expose other issues.
>
> The patch would also have to be re-worked to support all features in HDF5. It
has to support CMake, VDS, external links, etc..
>
> While the bug report has a high priority, it is not clear when the issue
can be addressed. It will take a lot of work to implement the patch.
>
> I sent a message to the developer and will let you know if he has any
further comments.

**Comment by tadeu** [comment]
Ok, thanks for the feedback!

We are working with a custom patch for 4 years and it seems to be fine: https://github.com/conda-forge/hdf5-feedstock/issues/47 

That patch is for an older version and it only the tackles the problem of Unicode in filenames (so that we are able to open filenames with unicode in it). It's Windows specific, so it doesn't necessarily need to use the POSIX layer.

**Comment by Guimoute** [comment]
If that helps anyone, I had issues with a file that I could not save or load because its parent folder contained a "µ".

Instead of loading `h5py.File(path, "r")` directly, I just have to do:

    import pathlib
    path = bytes(pathlib.WindowsPath(path))

**Comment by andreabedini** [comment]
> HDF5 latest version is 1.10.4, and there still seems to be no support for UTF-8 filenames. This is from current 1.11.3 development source:
> 
> https://bitbucket.hdfgroup.org/projects/HDFFV/repos/hdf5/browse/src/H5win32defs.h?at=d4234d0a98ff68eb55b3fa0d2246e61e53308e41#57
> 
> still using `_open` instead of `_wopen` in Windows :/

for those brave enough to keep looking at this, HDFGroup seems to have moved to GitHub and the incriminated line is 

https://github.com/HDFGroup/hdf5/blob/master/src/H5win32defs.h#L49

No mention of UTF in their issues though, perhaps someone could open an issue (not me, I haven't used HDF5 in years)

**Comment by tadeu** [comment]
Just to let you know that HDF5 1.10.6 is supporting UTF-8 filenames!
